# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/ipc/ContentChild.cpp
# Commit: 360b12c6b3e4
# Full Hash: 360b12c6b3e4e74203ba90324580a6c5e149b66b
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2019-07-16 04:07:00
# Regressor Bug: 1563604
# File Overlap Count: 1
# Description:
#   Bug 1563604 - Part 1: Confirm detach calls before dropping refs, r=peterv
#   
#   This ensures that the object is not discarded while there are still in-flight
#   messages from the parent process, and vice-versa, which could cause crashes if a
#   CC was run before any pending messages have arrived.
# ==============================================================================

diff -r 80fff55bee77 -r 360b12c6b3e4 dom/ipc/ContentChild.cpp
--- a/dom/ipc/ContentChild.cpp	Mon Jul 15 17:19:17 2019 +0000
+++ b/dom/ipc/ContentChild.cpp	Mon Jul 15 20:44:50 2019 +0000
@@ -3878,10 +3878,17 @@
 }
 
 mozilla::ipc::IPCResult ContentChild::RecvDetachBrowsingContext(
-    BrowsingContext* aContext) {
-  MOZ_RELEASE_ASSERT(aContext);
-
-  aContext->Detach(/* aFromIPC */ true);
+    uint64_t aContextId, DetachBrowsingContextResolver&& aResolve) {
+  // NOTE: Immediately resolve the promise, as we've received the message. This
+  // will allow the parent process to discard references to this BC.
+  aResolve(true);
+
+  // If we can't find a BrowsingContext with the given ID, it's already been
+  // collected and we can ignore the request.
+  RefPtr<BrowsingContext> context = BrowsingContext::Get(aContextId);
+  if (context) {
+    context->Detach(/* aFromIPC */ true);
+  }
 
   return IPC_OK();
 }